File: 2015/js/day02.js
import '../../utils/js/array'

const paperSize = ([l, w, h]) => {
  const base = 2 * l * w + 2 * w * h + 2 * h * l
  const areas = [l*w, w*h, h*l]
  return base + areas.min()
}

const bowSize = ([l, w, h]) => {
  const base = l*w*h
  const edges = [l+w, w+h, h+l]
  return base + edges.min()*2
}

const part = fn => boxes => boxes.map(fn).sum()
const part1 = boxes => boxes.apply(part(paperSize))
const part2 = boxes => boxes.apply(part(bowSize))

export { paperSize, bowSize, part1, part2 }

File: 2015/hs/test/Day02.hs
module Day02 where

type Box = [Int]

paperSize :: Box -> Int
paperSize [l, w, h] = area + minimum d
  where area = sum d * 2
        d = zipWith (*) [l, w, h] [w, h, l]

bowSize :: Box -> Int
bowSize [l, w, h] = vol + minimum edges * 2
  where vol = l*w*h
        edges = [l+w, w+h, h+l]

part1 :: [Box] -> Int
part1 box = sum $ map paperSize box

part2 :: [Box] -> Int
part2 box = sum $ map bowSize box